### PR TITLE
update state contract / fix tests, add schema

### PR DESCRIPTION
**Pull request by vmidyllic**
1. Fixed critical js vulnerabilities.
2. Added schema contract.
3. Added updated State contract.
4. Fixed tests.
5. Updated deploy migrations.
6. Updated Verifier contract according to the latest state
7. Added ci/cd tests

### CODE REVIEW

File: contracts/Schema.sol [hunk 1]
```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity >=0.6.0 <0.9.0;
+/**
+ * @title Schema
+ * @dev Schema contract
+ */
+contract SchemaRegistry {
+
+    struct Schema {
+        address issuer;
```
Comment (demonsh) [inline contracts/Schema.sol:10]: Probably better naming will be an owner or creator instead of the issuer? As it is an account that created the schema

Comment (vmidyllic) [reply contracts/Schema.sol:10]: the issuer of schema is an account that created this schema. It's ok

Comment (OBrezhniev) [reply contracts/Schema.sol:10]: Creator/owner makes here more sense than issuer for me too.

File: contracts/Schema.sol [hunk 2]
```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity >=0.6.0 <0.9.0;
```
Comment (OBrezhniev) [inline contracts/Schema.sol:2]: Probably version range should be more strict. What is the version of solidity compiler we are using?

File: contracts/Schema.sol [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    function getHashFromBytes(bytes memory schemaBody) private pure returns (bytes32) {
+        return keccak256(schemaBody);
+    }
+
+    /**
+     * @dev save is function to store schema
+     * @param schemaName - name of the schema
+     * @param schemaBody - json schema representation in bytes
+     */
+    function save(string memory schemaName, bytes memory schemaBody) public {
```
Comment (OBrezhniev) [inline contracts/Schema.sol:29]: How about `store` name for the function, anyway updating is not possible (maybe it should be?)

Comment (vmidyllic) [reply contracts/Schema.sol:29]: Save is ok. Update is not possible.

File: contracts/Schema.sol (hunk 2 above)
Comment (vmidyllic) [reply contracts/Schema.sol]: current truffle setup compiles all contracts with 0.6.0

Comment (vmidyllic) [reply contracts/Schema.sol]: changed to 0.6.0

File: contracts/Schema.sol (hunk 1 above)
Comment (vmidyllic) [reply contracts/Schema.sol]: fixed

File: contracts/Schema.sol (hunk 3 above)
Comment (demonsh) [reply contracts/Schema.sol:29]: save should be ok, as it is actually save operation

**Review by demonsh — APPROVED** [review]

File: contracts/State.sol [hunk 4]
```diff
@@ -7,212 +7,231 @@ import './lib/verifier.sol';
...
+        uint256 id,
+        uint[2] memory a,
+        uint[2][2] memory b,
+        uint[2] memory c
+    ) private {
+
+        uint256[3] memory input = [
+        id,
+        oldState,
+        newState
```
Comment (OBrezhniev) [inline contracts/State.sol:91]: not very good formatting here

Comment (OBrezhniev) [reply contracts/State.sol:91]: in general formatting of the file should be fixed

File: contracts/State.sol [hunk 5]
```diff
@@ -7,212 +7,231 @@ import './lib/verifier.sol';
...
+        ];
+        require(verifier.verifyProof(a, b, c, input), "zkProof idState update could not be verified");
+
+        identities[id].push(IDState(uint64(block.number), uint64(block.timestamp), newState));
+
+        transitions[oldState] = now;
```
Comment (OBrezhniev) [inline contracts/State.sol:97]: Probably there should be `newState`, not `oldState`, or variable name should be renamed.

Comment (OBrezhniev) [reply contracts/State.sol:97]: maybe `phasedOutAt` or `replacedAt`

Comment (vmidyllic) [reply contracts/State.sol:97]: transitions map stores timestamp when old state of identity transited to new state.

File: contracts/State.sol (hunk 4 above)
Comment (vmidyllic) [reply contracts/State.sol]: fixed with prettier

File: contracts/State.sol (hunk 5 above)
Comment (vmidyllic) [reply contracts/State.sol:103]: maybe name can be changed in next iteration but it will require integration changes

Comment (vmidyllic) [reply contracts/State.sol:103]: this mapping will be changed in the nearest feature, so we can take an appropriate name.

**Review by OBrezhniev — APPROVED** [review]

**Pull request merged**